File: Single_Number_II.cpp
class Solution {
public:
    /*
    O(1) space complexity
    Use a 32 size array to store the number of '1' s in every bit. If the ith bit has n_i number of '1's
    and n_i % 3 != 0, the single number should have 1 in ith bit.
    */
    int singleNumber(int A[], int n) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        if (n == 0) return n;         
        int counts[32] = {0};
        int result = 0;
        for(int i = 0; i < 32; i++){
            for(int j = 0; j < n; j++){
                if(((A[j] >> i) & 1) == 1){
                    counts[i] = (counts[i] + 1) % 3;
                }
            }
            result |= (counts[i] << i);
        }
        return result;
    }
};

File: Java/Single_Number_II.java
public class Solution {
    public int singleNumber(int[] A) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int n = A.length;
        if (n == 0) return 0;
        int one = 0, two = 0, zero = ~0;
        /*
        one: stores the bits with count % 3 = 1
        The ith bit of one is 1 indicates:
        Total number of appearances of 1s in the ith bit is k and k % 3 = 1
        The ith bit of one is 0 indicates:
        Total number of appearances of 1s in the ith bit is k and k % 3 != 1
        
        two: stores the bits with count % 3 = 2
        zero: stores the bits with count % 3 = 0
        */
        for (int i = 0; i < n; i ++) {
            two ^= (two & A[i]);   // clear the bits shifted up
            two |= (one & A[i]);   // include the bits shifted up from 'one'
            one ^= (one & A[i]);   // clear the bits shifted up
            one |= (zero & A[i]);  // include the bits shifted up from 'zero'
            zero = ~ (one | two);
        }
        return one;
    }
}
